### PR TITLE
feat(AvroBaseModel): non-deprecated pydantic v2 custom fields

### PR DESCRIPTION
**Pull request by fajpunk**
Add a new way to generate Avro schema and serialize custom fields for Pydantic v2 `AvroBaseModel`s. This new strategy provides a new `custom_encoders` `Meta` field that maps unsupported types to a new `CustomAvroEncoder` class. This class describes how to convert unsupported types to supported types.

The existing Pydantic v2 custom serialization/schema-generation strategy depends on the `json_encoders` model config, which is deprecated, and it doesn't seem to work with Unions (or maybe other container types).

My only usecase for this is so I can serialize timedeltas to float seconds in my models without having to depend on deprecated Pydantic functionality (and without having to write a custom __get_pydantic_core_schema__, because this is natively supported by Pydantic). If there's a better way to do this that I missed, please let me know (and thank you for your time)!

Another option would be to add native support for `timedelta`s, but there are a lot of different ways that folks might want to encode them.

### CODE REVIEW

**Comment by marcosschroh** [comment]
Hi @fajpunk 

Thanks for the PR. I think it makes a lot of sense, also now on `pydantic v2` it is quite hard to achieve the same functionality that we had with `pydantic v1`. I don't understand why we need to define [__get_pydantic_core_schema__](https://marcosschroh.github.io/dataclasses-avroschema/pydantic/#classes-with-__get_pydantic_core_schema__) which has a lot of boiler plate an it is complicated rather than [json_encoders](https://marcosschroh.github.io/dataclasses-avroschema/pydantic_v1/#classes-with-__get_validators__). 

The functionality that you propose is one that we miss, which I think is [duration](https://avro.apache.org/docs/1.12.0/specification/#duration) (timedelta). I think we should have support for it. This field should be a `datetime.timedelta` which is an `avro logicalType` represented as `bytes`. The representation to `json` and `dict` should be handled directly by pydantic and dataclasses. what do you think?

**Comment by fajpunk** [comment]
Yeah, the pydantic v2 `__get_pydantic_core_schema__` method definitely seems more complicated, even if it is more flexible. This PR makes it so that the custom encoding step is totally separate from any Pydantic functionality, which comes with some downsides. Some more thoughts on those downsides, and thoughts on built-in `timedelta` support.

TLDR:
* I'm glad to attempt native timedelta serialization in another PR if you can give me some guidance as to how to do the serialization to `duration` if you want that, given that `fastavro` doesn't support it
* I have some other thoughts on custom type serialization and I'm glad to try a different strategy with this PR if you want also

Thank you, as always, for this great lib, and for taking the time!


### `timedelta` support

I'd be happy to try to add native `timedelta` support in another PR! I agree that the representation to json and dict should be handled directly by pydantic and dataclasses.

#### What Avro type to serialize to?
I personally think that serializing to `double` seconds is more generally useful, and a better match for what `timedelta` in Python actually is. The main difference between a Python `timedelta` and an Avro `duration` is that a `timedelta` always represents the same amount of time, no matter where it is measured from, according to [the normalization logic](https://docs.python.org/3/library/datetime.html#datetime.timedelta). A day is always `3600 * 24` seconds, or `1000* 3600 * 24` milliseconds. In an Avro `duration`, the spec mentions explicitly that a day can have a different number of milliseconds depending on when it is measured from:

```...depending on the moment in time from which the duration is measured, the number of days in the month and number of milliseconds in a day may differ```

But, I think most people disagree with me on this :smile: Other Pydantic -> Avro schema libs have chosen `duration`, for example:
* [pydantic2avro](https://github.com/Happy-Kunal/pydantic2avro/blob/1007270425e831cc19c40de72ac54c65675906b8/src/pydantic2avro/schema_maker.py#L120-L121)
* [py-avro-schema](https://github.com/jpmorganchase/py-avro-schema/blob/a68745c85420bea5e34883ab9671366441efa6e8/src/py_avro_schema/_schemas.py#L464-L471)

Pydantic itself [exposes an config option](https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.ser_json_timedelta) to serialize to either `float` seconds, or an [ISO 8601 Duration](https://en.wikipedia.org/wiki/ISO_8601#Durations). An ISO 8601 Duration is ambiguous in the same way that an Avro `duration` is, but I like that Pydantic gives you the option.

What do you think about doing something similar to Pydantic and adding a `Meta` class [option](https://marcosschroh.github.io/dataclasses-avroschema/records/#class-meta) so folks can decide between `double` seconds and `duration` (assuming we can figure out how to de/serialize `duration`s)?

#### Implementation

* Whichever way you think is best, I think I can see how to implement the schema generation. But I'm not sure how offhand how to implement de/serialization to and from `duration`, given that [fastavro doesn't support it](https://fastavro.readthedocs.io/en/latest/logical_types.html?highlight=duration).

### Custom type handling
This PR implements Avro schema generation and encoding of custom Python types to `fastavro`-supported python types, for all model types: dataclasses, Pydantic v1, and Pydantic v2. But it leaves it up to the model to know how to get from the fastavro-deserialized Python type back to the custom Python type. In other words, the encoding happens in functionality specific to this library, but the decoding needs to happen outside of this library, which I admit is confusing :frown:. With `timedelta`, for example, I couldn't find a clean way to figure out just from the Avro data whether a `double` needed to be deserialized back into a `timedelta` rather than a float. I think the schema generation would need to add some metadata.

My first attempt was to implement this for just Pydantic v2 models, using the native Pydantic [serialization](https://docs.pydantic.dev/latest/concepts/serialization/) and [validation](https://docs.pydantic.dev/latest/concepts/validators/) `Annotated` functionality. But I ran into a couple of problems:
* We could potentially use [serialization context](https://docs.pydantic.dev/latest/concepts/serialization/#serialization-context) and [validation context](https://docs.pydantic.dev/latest/concepts/validators/#validation-context), and pass something that folks could use in their functions that lets them know that we're trying to de/serialize to Avro, but that depends on using `model_dump`, and currently, [pydantic models are transformed with `dict`](https://github.com/fajpunk/dataclasses-avroschema/blob/custom-serializers-take-2/dataclasses_avroschema/pydantic/main.py#L36-L36), which doesn't allow context to be passed in. This seems like it can be changed maybe?
* Defaults in Avro schema generation: I couldn't figure out how to serialize a single value for a field through all of the Pydantic machinery (instead of serializing an entire model instance). This seems like a harder problem.

**Comment by marcosschroh** [comment]
Hey, after giving it some thoughts I came with some ideas:

1. At his moment we do not have support for the native `timedelta`  python type, so adding it would be a great addition. I don't know whether it should be the same as `avro duration`, plus as you mentioned it seems not supported by `fastavro`, so it is tricky. If we have a new field type `TimeDeltaField` then it is a win because it will work for `dataclasses`, `pydantic v1 and v2` and `faust`. The avro representation of this new field should be `float`, we need to generate proper `fakes` and check the `default`. Then, the function [standardize_custom_type](https://github.com/marcosschroh/dataclasses-avroschema/blob/master/dataclasses_avroschema/utils.py#L95) should be updated to handle `timedelta`.

2. I think having `customer_encoders` it can have wider benefits, for example the introduction of `customer logical types` or handle cases when the `AvroModel` has a field which a type is a class that comes from third packages and it is a bit tricky to make it work because they do not inherit from `AvroModel`.

3. What happens with `model generation` when using `customer_encoders`? Following your docs examples, from a model is ` timedelta` which is translated to a `float`, but when going back we get a `float` rather than a `timedelta` type.

I think that it could be great to have a separate `PR` to add the `TimeDeltaField` which:

-  It will behave as `float` in `avro`
- We should generate proper fakes
- We must adapt the `standardize_custom_type` funtion
- Make sure that this new field type will work with all the flavors (dataclasses, faust and pydanti)
- Introduce `metadata` when generating the schema so we can use it when go `schema` -> `model`, something similar to https://marcosschroh.github.io/dataclasses-avroschema/pydantic/#model-generation

Then, we should continue with the `custom_encoders` keeping in mind:

-  Custom logical types
- How to handle custom types that do not inherit from `AvroModel`
- How to generate the `custom_encoders` entry in the `class Meta` when  `schema` -> `model`

Thoughts ?

**Comment by fajpunk** [comment]
Ok! I'll get started on another PR for `TimeDeltaField`, then I can come back to this one if you want.

> I think having customer_encoders it can have wider benefits, for example the introduction of customer logical types or handle cases when the AvroModel has a field which a type is a class that comes from third packages and it is a bit tricky to make it work because they do not inherit from AvroModel.

If I'm understanding correctly, this works the same way as the current `json_encoders` implementation. `custom_encoders` is sort of a map from `unsupported python type -> supported python type`.  So as long as there's an entry in `Meta.custom_encoders` on the parent `AvroModel` class, the value of any field of the custom type will be converted to the supported type in the generated Avro schema and when serializing to Avro.

> What happens with model generation when using customer_encoders? Following your docs examples, from a model is  timedelta which is translated to a float, but when going back we get a float rather than a timedelta type.

Yep, it doesn't work :frowning_face: Model generation is unsupported, just like with `json_encoders`, but it should probably blow up rather than silently doing the wrong thing.

**Comment by fajpunk** [comment]
Closing this for now, since a better solution will likely be very different.
